aya/src/programs/mod.rs

```diff
@@ -216,6 +218,10 @@ pub enum ProgramError {
     /// An error occurred while working with IO.
     #[error(transparent)]
     IOError(#[from] io::Error),
+
+    /// A checked integral type conversion failed.
+    #[error(transparent)]
+    TryFromIntError(#[from] TryFromIntError),
```

Comment:
We shouldn't do this. We're exposing an implementation detail to the public
API: what if we stop using try_from/try_into in the implementation? Or if I get
the error, I get "TryFromIntError" with an error "an integral type conversion
failed" - what? what did I do wrong? how do I fix it?

Where is this used from? If we must fail, we should use a better, higher level
error variant.

Reply:
I’ve made the error more descriptive.

I’m only using it in cases where an if index is being converted from `u32` to `i32`.  The problem stems from the fact that the kernel isn’t very disciplined about whether to use a `u32` or `i32` for if index, so we need to convert sometimes.  The existing code was using the unchecked `as i32` to do the conversion, but Tamir wanted that changed to a checked `try_into`.

What happens if it fails?  In these cases, it should never fail because the value should have started out as an `i32` in the kernel, but got converted to a `u32` somewhere before we got it, so it should convert back to an `i32` just fine.  In the unlikely event that it does fail, there’s a bug either in our code or in one of the external Linux functions, and the user would need to report it and/or fix it.